app/celery/signals.py

```diff
@@ -0,0 +1,103 @@
+import inspect
+import asyncio
+from celery.signals import (
+    task_received, task_internal_error, task_revoked,
+    task_success, task_failure, task_retry, task_prerun
+)
+from celery_singleton.util import generate_lock
+from celery_singleton.config import Config
+from celery_singleton.backends import get_backend
+from app.celery.worker import celery_app
+from app.database import SessionLocal
+from app.services import WebsocketManagerService, JobStatusService, InstructorService
+from app.models.job_status import JobStatusModel, JobStatus
+from app.schemas import JobSchema, WebsocketJobStatusMessage
+
+def handle_job_status_update(job_id: str, job_type: str | None, job_status: JobStatus) -> None:
+    """ Store a job status update in the database and dispatch a websocket event. """
+    async def _handle_job_status_update():
+        with SessionLocal() as session:
+            job_status_model = JobStatusService(session).create_job_status_update(job_id, job_type, job_status)
+            ws_message = WebsocketJobStatusMessage(
+                data=JobSchema.from_orm(job_status_model)
+            )
+            
+            if job_type == "downsync":
+                user_list = await InstructorService(session).list_instructors()
+            else:
+                user_list = []
+
+            WebsocketManagerService.publish_sync_pubsub_ws_message(ws_message, user_list)
+    
+    asyncio.run(_handle_job_status_update())
```

Comment:
What's the difference between this and just making the `handle_job_status_update()` method async? In the way you have it you're kind of disguising the fact that the internals of this function are async, which I don't love.

Reply:
It has been so long since I've done these changes that I can't say this will be accurate with 100% confidence, but I'm fairly certain I had to write it in such a way due to the way asynchronous code and the event loop is handled on Celery workers, where signals are triggered. If you did it the typical way (that you would expect to see here) it would cause errors related to asyncio.